[FLINK-12847] [Connectors / Kinesis] update flink-connector-kinesis to use Apache 2.0 licensed code

dyanarose opened the pull request:
## What is the purpose of the change

The Kinesis Connector will now be able to be built and included in the build artefacts as it no longer pulls in any Amazon licensed code.

## Brief change log
 - update Kinesis Client Library to v1.11.2
 - update Kinesis Producer Library to v0.13.1
 - update DynamoDBStreams Library to v1.5.0
 - update NOTICE files
 - update pom and travis/stage.sh to remove the include-kinesis profile and to add the flink-connector-kinesis module to the modules list
 - update documentation for the kinesis connector

## Verifying this change
This change is already covered by existing tests, such as the full suite of tests for the kinesis connector.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

bowenli86 commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -12,15 +12,11 @@ This project bundles the following dependencies under the Apache Software Licens
 - com.amazonaws:jmespath-java:1.11.319
 - org.apache.httpcomponents:httpclient:4.5.3
 - org.apache.httpcomponents:httpcore:4.4.6
+- com.amazonaws:amazon-kinesis-client:1.10.0
```
my "mvn dependency:tree" shows

```
[INFO] +- com.amazonaws:amazon-kinesis-client:jar:1.11.2:compile
[INFO] |  +- com.amazonaws:aws-java-sdk-dynamodb:jar:1.11.603:compile
[INFO] |  |  \- com.amazonaws:aws-java-sdk-s3:jar:1.11.603:compile
[INFO] |  |     \- com.amazonaws:aws-java-sdk-kms:jar:1.11.603:compile
[INFO] |  \- commons-logging:commons-logging:jar:1.1.3:compile
```

thus, are we missing `aws-java-sdk-dynamodb`, `aws-java-sdk-s3`, and `aws-java-sdk-kms`?

dyanarose replied:
most likely, there were a few sub-dependencies missing in the notices and i was unsure of where they should be listed. I'll add these in here.

bowenli86 commented:
cc @zentol @gorde

dyanarose commented on flink-connectors/flink-connector-kinesis/pom.xml:
```diff
@@ -34,9 +34,9 @@ under the License.
-		<aws.kinesis-kcl.version>1.9.0</aws.kinesis-kcl.version>
-		<aws.kinesis-kpl.version>0.12.9</aws.kinesis-kpl.version>
-		<aws.dynamodbstreams-kinesis-adapter.version>1.4.0</aws.dynamodbstreams-kinesis-adapter.version>
+		<aws.kinesis-kcl.version>1.10.0</aws.kinesis-kcl.version>
```
nah, just hadn't seen 1.11.x was now out. updating

dyanarose commented:
I've updated the version of the KCL to v1.11.2. I've updated the version of the AWS SDK to be consistent. 

The NOTICE file should now have all compiled dependencies. And I've, hopefully, fixed the hash I made of the NOTICE-binary file and added the kinesis-connector in to the correct location 🤕.

zentol commented on .travis.yml:
```diff
@@ -260,91 +260,86 @@ jobs:
       script: ./tools/travis/nightly.sh split_misc_hadoopfree.sh
       name: misc
     - if: type = cron
-      env: PROFILE="-Dinclude-kinesis"
```
set this to an empty string instead; otherwise the PROFILE value of the above job will be used. (the travis.yml is funky like that.........)

zentol commented on NOTICE-binary:
```diff
@@ -54,6 +54,40 @@ Copyright 2006-2019 The Apache Software Foundation
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
 
+flink-connector-kinesis
```
kinesis isn't bundled in the distribution, and hence this stuff can be removed.

zentol commented on docs/dev/connectors/kinesis.zh.md:
```diff
@@ -38,13 +38,13 @@ To use the connector, add the following Maven dependency to your project:
 
-The `flink-connector-kinesis{{ site.scala_version_suffix }}` artifact is not deployed to Maven central as part of
-Flink releases because of the licensing issue. Therefore, you need to build the connector yourself from the source.
+<span class="label label-danger">Attention</span> Prior to Flink version 1.10.0 the `flink-connector-kinesis{{ site.scala_version_suffix }}` has a dependency on code licensed under the [Amazon Software License](https://aws.amazon.com/asl/) (ASL).
```
let's not shorting this to ASL; it just invites ambiguity.

zentol commented on docs/dev/connectors/kinesis.zh.md:
```diff
@@ -38,13 +38,13 @@ To use the connector, add the following Maven dependency to your project:
-The `flink-connector-kinesis{{ site.scala_version_suffix }}` artifact is not deployed to Maven central as part of
-Flink releases because of the licensing issue. Therefore, you need to build the connector yourself from the source.
+<span class="label label-danger">Attention</span> Prior to Flink version 1.10.0 the `flink-connector-kinesis{{ site.scala_version_suffix }}` has a dependency on code licensed under the [Amazon Software License](https://aws.amazon.com/asl/) (ASL).
+Linking to the flink-connector-kinesis for the prior versions will include ASL licensed code into your application.
```
```suggestion
Linking to the prior versions of flink-connector-kinesis will include this code into your application.
```

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- com.fasterxml.jackson.core:jackson-databind:2.6.7.2
+- com.fasterxml.jackson.core:jackson-core:2.6.7
+- com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:2.6.7
+- com.google.guava:guava:18.0
```
this is not bundled in the jar

dyanarose replied:
ta, wasn't certain on these, so went searching to see if other NOTICE files referenced them.

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- com.amazonaws:aws-java-sdk-s3:1.11.603
+- com.amazonaws:aws-java-sdk-sts:1.11.603
+- com.amazonaws:jmespath-java:1.11.603
+- com.fasterxml.jackson.core:jackson-annotations:2.6.0
```
jackson as a whole is is not bundled in the jar

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- com.google.guava:guava:18.0
+- commons-codec:commons-codec:1.10
+- commons-lang:commons-lang:2.6
+- commons-logging:commons-logging:1.1.3
```
this is not bundled in the jar

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:2.6.7
+- com.google.guava:guava:18.0
+- commons-codec:commons-codec:1.10
+- commons-lang:commons-lang:2.6
```
this is not bundled in the jar

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- commons-codec:commons-codec:1.10
+- commons-lang:commons-lang:2.6
+- commons-logging:commons-logging:1.1.3
+- joda-time:joda-time:2.5
```
this is not bundled in the jar

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,29 @@ The Apache Software Foundation (http://www.apache.org/).
+- com.fasterxml.jackson.core:jackson-core:2.6.7
+- com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:2.6.7
+- com.google.guava:guava:18.0
+- commons-codec:commons-codec:1.10
```
this is not bundled in the jar

zentol commented:
The kinesis E2E test fails by the way.

zentol commented:
```2019-08-22 08:44:33,704 WARN  org.apache.flink.kinesis.shaded.com.amazonaws.http.apache.utils.ApacheUtils  - NoSuchMethodError was thrown when disabling normalizeUri. This indicates you are using an old version (< 4.5.8) of Apache http client. It is recommended to use http client version >= 4.5.9 to avoid the breaking change introduced in apache client 4.5.7 and the latency in exception handling. See https://github.com/aws/aws-sdk-java/issues/1919 for more information```

dyanarose commented:
> `2019-08-22 08:44:33,704 WARN org.apache.flink.kinesis.shaded.com.amazonaws.http.apache.utils.ApacheUtils - NoSuchMethodError was thrown when disabling normalizeUri. This indicates you are using an old version (< 4.5.8) of Apache http client. It is recommended to use http client version >= 4.5.9 to avoid the breaking change introduced in apache client 4.5.7 and the latency in exception handling. See https://github.com/aws/aws-sdk-java/issues/1919 for more information`

🤔 hmmm:
 com.amazonaws:aws-java-sdk-kinesis:jar:1.11.603:compile
[INFO] |  +- com.amazonaws:aws-java-sdk-core:jar:1.11.603:compile
[INFO] |  |  +- org.apache.httpcomponents:httpclient:jar:4.5.3:compile
[INFO] |  |  |  +- org.apache.httpcomponents:httpcore:jar:4.4.6:compile


I'll have a check to see why it's pulling in these versions

dyanarose commented on flink-connectors/flink-connector-kinesis/pom.xml:
```diff
@@ -151,6 +147,18 @@ under the License.
+		<dependencies>
+
+			<dependency>
+				<groupId>org.apache.httpcomponents</groupId>
```
4.5.3 is specifically being pulled in in the parent pom, and com.amazonaws:aws-java-sdk-kinesis only specifies a soft version requirement on 4.5.9, so looks to me like 4.5.3 is being used as it's in scope.

tweise replied:
Perhaps as follow-up (outside this PR) upgrade the dependency in the parent pom and then we can eliminate this here.

dyanarose replied:
i'll see about popping an issue into jira for it and link to it's declaration in the parent pom.

dyanarose commented on NOTICE-binary:
```diff
@@ -54,7 +54,6 @@ Copyright 2006-2019 The Apache Software Foundation
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
 
-
```
just pegging this in here, to make sure we're all on the same page: https://github.com/apache/flink/pull/9494#discussion_r316597961

Since the kinesis connector will be building and released with Flink once this PR is merged in, we should be adding the dependencies to the NOTICE-binary file, correct?

@zentol

dyanarose commented:
> Please also update documentations upon removing "include-kinesis"
> 
> ```
> $ ag include-kinesis
> dev/connectors/kinesis.zh.md
> 49:mvn clean install -Pinclude-kinesis -DskipTests
> 52:mvn clean install -Pinclude-kinesis -DskipTests
> 62:mvn clean install -Pinclude-kinesis -Daws.kinesis-kpl.version=0.12.6 -DskipTests
> 
> dev/connectors/kinesis.md
> 48:mvn clean install -Pinclude-kinesis -DskipTests
> 51:mvn clean install -Pinclude-kinesis -DskipTests
> 61:mvn clean install -Pinclude-kinesis -Daws.kinesis-kpl.version=0.12.6 -DskipTests
> ```

these are referring to how to build for prior versions.   

🤔 should this entire section be removed do you think. Leave in the mention of the licensing issue in prior versions, and then say something like "refer to the kinesis connector documentation for the version you are running" ?

bowenli86 commented:
> @bowenli86
> 
> these are referring to how to build for prior versions.
> 
> 🤔 should this entire section be removed do you think. Leave in the mention of the licensing issue in prior versions, and then say something like "refer to the kinesis connector documentation for the version you are running" ?

yeah, let's remove them. Flink has versioned docs with each release, users using an old kinesis connector should reference docs of a previous version

bowenli86 commented:
@dyanarose can you rebase to latest master? Otherwise, LGTM.

@zentol @tweise if you don't have other concerns , I'll merge it after the rebase and build passes

zentol commented:
I'll try to make another pass later today.

zentol commented on flink-connectors/flink-connector-kinesis/src/main/resources/META-INF/NOTICE:
```diff
@@ -6,21 +6,19 @@ The Apache Software Foundation (http://www.apache.org/).
-- org.apache.httpcomponents:httpclient:4.5.3
+- com.amazonaws:amazon-kinesis-client:1.11.2
+- com.amazonaws:amazon-kinesis-producer:0.13.1
+- com.amazonaws:dynamodb-streams-kinesis-adapter:1.5.0
```
please move this one below the remaining `aws-java` entries so the list is sorted

dyanarose replied:
👍 I assume no one will mind if I squash before pushing this change, to avoid another rebase after

tweise commented:
@dyanarose thanks for working in this! Minor: please reformat the commit message:
```
[FLINK-12847][kinesis] update flink-connector-kinesis to use Apache 2.0 license code
```

bowenli86 commented:
Merging
